### PR TITLE
Added draft of a new documentation quality policy.

### PR DESCRIPTION
**Pull request by ulrikeyang**
Hi all,
I am opening a pull request to add a policy on documentation quality. Please take a look and provide your comments.
Thanks
Ulrike

### CODE REVIEW

File: package_policies/R8.md [hunk 1]
```diff
@@ -0,0 +1 @@
+**R8.** Each xSDK member package should have sufficient documentation to support use and further development. Elements may include, for example, getting started, functionality documentation, on-ramping new developers, examples, and team processes. Sufficiency of documentation will be reviewed by the xSDK community.
```
Comment (bangerth) [inline package_policies/R8.md]:
```suggestion
**R8.** Each xSDK member package should have sufficient documentation to support use and further development.
While it is difficult to formalize what constitutes sufficiency of documentation, the requirement will be reviewed by the
xSDK community based on documentation elements that may include, for example, the following:
  * Instructions on installation and getting started;
  * Documentation on functionality, for example class- and member-level documentation (in object-oriented languages) or more generally API-level documentation;
  * Discussions of how individual components of the code base interact (for example, in object-oriented languages, how classes or groups of classes interact), ideally using "typical use case" code snippets;
  * Presence of complete example codes that illustrate typical use cases;
  * Links to external resources such as mailing lists and forums where users and developers can find help;
  * Instructions for on-ramping new developers;
  * Description of team processes.
```

**Review by bangerth — COMMENTED** [review]
I expanded this slightly. I don't know whether you want to break it into a bulleted list, but I thought this is at least a good starting point for further discussions.

File: package_policies/R8.md (hunk 1 above)
Comment (jedbrown) [reply package_policies/R8.md]: + Comparison of strengths and weaknesses relative to related software, to assist a prospective user in choosing appropriate technology.
+ Theory manual where appropriate, laying out the relevant mathematical abstractions and notation.
+ Low-commitment demos using technologies such as Binder or containers.

Comment (hartwiganzt) [reply package_policies/R8.md]: I think @bangerth slightly longer version clarifies the quality metrics and helps the softwaredevelopers in deciding what documenttion to provide.

Comment (ulrikeyang) [reply package_policies/R8.md]: Comments from the xSDK meeting on July 23:
We discussed Wolfgang's suggestion. Here are a few questions that came up:
Ideal: What is the "good enough" for a new user to get started?
Want to isolate the most important components for the mandatory policy (down the road). 
Instructions for installation and "Getting started" (this is still subjective). 
Careful that language isn't too C++ specific: E.g. member and class, maybe use 'API' instead.
Make language as generic as possible (Applicable to python, c, c++)

**Comment by osborn9** [comment]
General comments from 7/23 xSDK meeting: 
- Ideal: Minimal set of "good enough" set of information for a new user to get started.
- Isolate the most important components for the mandatory policy (down the road). 
    * Instructions for installation
    * "Getting started" (this is still subjective -- level of detail)
- Careful that language isn't too C++ specific: E.g. member and class
    * Use 'API' instead 
    * Make language as generic as possible (Applicable to python, c, c++)

File: package_policies/R8.md (hunk 1 above)
Comment (bangerth) [reply package_policies/R8.md]: Apologies for not being on the call yesterday -- I was on vacation this week, while also participating in two workshops; something had to give :-)

Do you want me to edit anything in my comments above, or did you come to agreement?

**Comment by ulrikeyang** [comment]
Hi Wolfgang
I completely understand.  I didn’t expect everybody to attend knowing about the conflict with Collegeville.There has been no agreement yet, I.e. we will continue the discussion at the next meeting, but there has been progress. People generally like your suggestion, but would like it modified, less C++ specific, since we also have C and python libraries. Feel free to make suggestions or you can wait until next time. I hope you were able to vacation a bit at least.


---
Sent from Workspace ONE Boxer<https://whatisworkspaceone.com/boxer>

On July 24, 2020 at 7:35:33 PM PDT, Wolfgang Bangerth <notifications@github.com> wrote:

@bangerth commented on this pull request.

________________________________

In package_policies/R8.md<https://github.com/xsdk-project/xsdk-community-policies/pull/22#discussion_r460353795>:

> @@ -0,0 +1 @@
+**R8.** Each xSDK member package should have sufficient documentation to support use and further development. Elements may include, for example, getting started, functionality documentation, on-ramping new developers, examples, and team processes. Sufficiency of documentation will be reviewed by the xSDK community.


Apologies for not being on the call yesterday -- I was on vacation this week, while also participating in two workshops; something had to give :-)

Do you want me to edit anything in my comments above, or did you come to agreement?

**Comment by bangerth** [comment]
I will (again) not be able to join today, but have updated the change suggestions in the comment above (https://github.com/xsdk-project/xsdk-community-policies/pull/22#pullrequestreview-430861554) to read less C++-specific. 

(I do think that talking about classes and members has little potential for confusion because essentially every scientific software library/package I know of is "object oriented" whether it is written in C++/Python or C/Fortran: PETSc is a good example in this as it is perfectly object oriented even though it is written in C. I think that the principles of OO software design have diffused in our community regardless of the languages used, and what is a "class" and a "member" is probably sufficiently obvious that most developers can map that to whatever language construct they use in expressed OO principles in their language.)

**Comment by jedbrown** [comment]
I tend to agree that OO language is okay; we can entrust developers to reasonably extrapolate the intent to Rust traits or whatever is appropriate terminology in their chosen language/paradigm.

File: package_policies/R8.md (hunk 1 above)
Comment (mkstoyanov) [reply package_policies/R8.md]: We can have two sections, recommended and strongly recommended:

**Strongly recommended:**
* Instruction on installation and getting started, e.g., list of relevant CMake or Autoconf options
* Documentation of the front-facing (external) API, e.g., structs, classes, methods, etc.
* Provide a set of examples of common use-cases

**Recommended:**
* Documentation of the internal API, e.g., methods and classes that are not going to be directly called by the user
* Overview of the internal design describing the relations between different components
* Description of the pull-request policies, e.g., formatting and naming conventions, C++ standard, etc.
* Theory manual where appropriate, laying out the relevant mathematical abstractions and notation.
* Comparison of strengths and weaknesses relative to related software, to assist a prospective user in choosing appropriate technology.
* Low-commitment demos using technologies such as Binder or containers.

The last three are directly copied from @jedbrown 

Eventually the top three should be come mandatory.

Comment (hartwiganzt) [reply package_policies/R8.md]: I am very happy with the current version. One comment maybe: I think the term "mailing list" should be understood in a broad sense and include modern versions of "mailing lists" such as Slack, Mattermost etc.

Comment (bangerth) [reply package_policies/R8.md]: Right. I think the proposed changes above, saying "mailing lists and forums" cover this.

**Review by balos1 — CHANGES_REQUESTED** [review]
A short summary of the policy should be added to https://github.com/xsdk-project/xsdk-community-policies/blob/master/package_policies/README.md before merging.

**Comment by ulrikeyang** [comment]
I added the short description to README and merged the new policy into the main.

**Pull request merged**